### ISSUE TITLE
Null pointer dereference Vulnerability in OpenSSL Function Structure Loading

### ISSUE THREAD

**Issue by xc01** [openssl/openssl#22950]
### Summary
Insufficient memory during program runtime can result in multiple null pointer returns from malloc, leading to a segmentation fault.

### Version
Discovered version:  3.0.9

### Description
In crypto/context.c, function context_init is used to initialize default_context_int, and it is globally utilized only once. If malloc failure occurs (due to restrcted memory or fault injection), the program will goto err tag and all fields in default_context_int will be set to NULL. It will trigger a fault later.
```
 err:
    if (exdata_done)
        ossl_crypto_cleanup_all_ex_data_int(ctx);
    for (i = 0; i < OSSL_LIB_CTX_MAX_INDEXES; i++)
        CRYPTO_THREAD_lock_free(ctx->index_locks[i]);
    CRYPTO_THREAD_lock_free(ctx->oncelock);
    CRYPTO_THREAD_lock_free(ctx->lock);
    memset(ctx, '\0', sizeof(*ctx));
    return 0;
```

In lhash.c/OPENSSL_LH_insert(), when the getrn() function detects that a function (which is used to define modules such as x509) from the data is not currently stored, it uses OPENSSL_malloc() to allocate space and store the new Function structure.

```
void *OPENSSL_LH_insert(OPENSSL_LHASH *lh, void *data)
{
    unsigned long hash;
    OPENSSL_LH_NODE *nn, **rn;
    void *ret;

    lh->error = 0;
    if ((lh->up_load <= (lh->num_items * LH_LOAD_MULT / lh->num_nodes)) && !expand(lh))
        return NULL;        /* 'lh->error++' already done in 'expand' */

    rn = getrn(lh, data, &hash);
```

If malloc encounters insufficient space and returns NULL, in openssl.c/do_cmd(), lh_FUNCTION_retrieve() will return NULL due to the inability to find the module. When fp is NULL, it will enter EVP_get_digestbyname, forcing the search or creation of a Function type structure.

```
    fp = lh_FUNCTION_retrieve(prog, &f);
    if (fp == NULL) {
        if (EVP_get_digestbyname(argv[0])) {
            f.type = FT_md;
            f.func = dgst_main;
            fp = &f;
        } else if (EVP_get_cipherbyname(argv[0])) {
            f.type = FT_cipher;
            f.func = enc_main;
            fp = &f;
        }
    }
```

Eventually, in context.c, ossl_lib_ctx_get_data() will be called, and it will use ossl_lib_ctx_get_concrete to obtain the previously mentioned default_context_int. As mentioned earlier, ctx->lock is NULL, so threads_pthread.c/CRYPTO_THREAD_read_lock() will lead to a crash due to dereferencing a null pointer.

```
OSSL_LIB_CTX *ossl_lib_ctx_get_concrete(OSSL_LIB_CTX *ctx)
{
#ifndef FIPS_MODULE
    if (ctx == NULL)
        return get_default_context();
#endif
    return ctx;
}
...
static OSSL_LIB_CTX *get_default_context(void)
{
    OSSL_LIB_CTX *current_defctx = get_thread_default_context();

    if (current_defctx == NULL)
        current_defctx = &default_context_int;
    return current_defctx;
}
...
__owur int CRYPTO_THREAD_read_lock(CRYPTO_RWLOCK *lock)
{
# ifdef USE_RWLOCK
    if (pthread_rwlock_rdlock(lock) != 0)
        return 0;
# else
    if (pthread_mutex_lock(lock) != 0) {
        assert(errno != EDEADLK && errno != EBUSY);
        return 0;
    }
# endif

    return 1;
}
```

**Comment by nhorman** [comment]
Theres only going to be so much to be done in low memory conditions as you describe:

It may be that the best thing to do is add a check in get_default context for default_context_inited == 1.  If that fails we could abort or perhaps call an optionally registered callback to make the application aware of the fact that there is no memory available.  But in the end, and out of memory condition is largely unrecoverable, at least not without application assistance to restart the initialization process.
